Sources/CollectionDiffing.swift

```diff
@@ -0,0 +1,560 @@
+import Foundation
+import Result
+
+/// Represents a snapshot of a collection.
+///
+/// The changeset associated with the first snapshot received by any given observation
+/// is generally irrelevant. Observations should special case the first snapshot as a
+/// complete replacement, or whatever semantic that fits their purpose.
+///
+/// `Snapshot` does not include a reference to the previous snapshot. If the
+/// history is relevant for delta consumption, it is up to the observers to cache the
+/// history on their own.
+///
+/// While the protocol does not bind the `Changeset` associated type with any constraint,
+/// it is expected to be either a `Changeset`, or a composite of `Changeset`s for a nested
+/// collection.
+public struct Snapshot<Elements: Collection, Changeset>: SnapshotProtocol {
+	public let elements: Elements
+	public let changeset: Changeset
+
+	public init(elements: Elements, changeset: Changeset) {
+		(self.elements, self.changeset) = (elements, changeset)
+	}
+}
+
+public protocol SnapshotProtocol {
+	associatedtype Elements: Collection
+	associatedtype Changeset
+
+	var elements: Elements { get }
+	var changeset: Changeset { get }
+}
+
+/// Represents an atomic batch of changes made to a collection.
+///
+/// A `Changeset` represents changes as **offsets** of elements. You may
+/// subscript a collection of zero-based indexing with the offsets, e.g. `Array`. You must
+/// otherwise convert the offsets into indices before subscripting.
+public struct Changeset {
+	/// Represents the context of a move operation applied to a collection.
+	public struct Move {
+		public let source: Int
+		public let isMutated: Bool
+
+		public init(source: Int, isMutated: Bool) {
+			(self.source, self.isMutated) = (source, isMutated)
+		}
+	}
```

Comment:
Any particular reason for choosing this representation?

The other possibility is to do something like this:

```swift
public enum Change {
    case insert(Int)
    case remove(Int)
    case move(from: Int, to: Int, mutated: Bool)
    case mutate(Int)
}
```

Not saying one is better than the other. But curious as to your rationale.